File: e2e/playwright-typescript/tests/edit-task/edit-task.spec.ts
import { test } from '@/fixtures';
import { TaskContext } from '@/context/TaskContext';
import { TaskPriority } from '@/data/TaskPriority';
import { TaskStatus } from '@/data/TaskStatus';
import * as allure from "allure-js-commons";
import { AllureEpic } from '@/data/AllureEpic';

test.beforeAll(async () => {
  await allure.epic(AllureEpic.TASK_MANAGEMENT);
  await allure.feature('Edit task');
  await allure.tms('101');
});

test.describe('Edit task', () => {
  let context: TaskContext;

  test.beforeEach(async ({ support }) => {
    context = new TaskContext({ status: TaskStatus.TODO, priority: TaskPriority.MEDIUM });
    const response = context.createTaskResponse();
    await support.mock.api.getTasks([response]);
    await support.mock.api.getTask(response.id, response);
    await support.mock.api.getIsValid(response.id, true);
  });

  test('should edit task', async ({ step, validate, support }) => {
    // given
    const updatedContext = new TaskContext({
      id: context.id,
      title: context.title + '-Updated',
      description: context.description + '-Updated',
      status: TaskStatus.IN_PROGRESS,
      priority: TaskPriority.HIGH,
    });

    const updatedResponse = updatedContext.createTaskResponse();

    await step.navigation.openMainPage();

    // when
    await step.tasks.openEditTaskForm(context.title);
    await step.tasks.editTask.fillForm(updatedContext.createTaskData());

    await support.mock.api.updateTask(updatedContext.id, updatedResponse);
    await support.mock.api.getTasks([updatedResponse]);
    await support.mock.api.getTask(updatedContext.id, updatedResponse);
    await support.mock.api.getIsValid(updatedContext.id, true);

    await step.tasks.editTask.submitForm();
    await step.tasks.openTaskInfoForm(updatedContext.title);

    // then
    await validate.task.data(updatedContext.createTaskData());
  });
});

File: e2e/capybara-ruby/spec/edit_task/edit_task_spec.rb
# frozen_string_literal: true

require 'spec_helper'

require_relative '../../context/task_context'
require_relative '../../data/allure_epic'
require_relative '../../data/task_priority'
require_relative '../../data/task_status'

RSpec.describe 'Edit task', type: :system,
                            epic: AllureEpic::TASK_MANAGEMENT,
                            feature: 'Edit task',
                            tms: '101' do
  let(:context) do
    TaskContext.new(status: TaskStatus::TODO, priority: TaskPriority::MEDIUM)
  end

  before do
    response = context.create_task_response
    support.mock.api
         .get_tasks([response])
         .get_task(response[:id], response)
         .get_is_valid(response[:id], true)
  end

  it 'should edit task' do
    # given
    updated_context = TaskContext.new(
      id: context.id,
      title: "#{context.title}-Updated",
      description: "#{context.description}-Updated",
      status: TaskStatus::IN_PROGRESS,
      priority: TaskPriority::HIGH
    )
    updated_response = updated_context.create_task_response

    step.navigation.open_main_page

    # when
    step.tasks.open_edit_task_form(context.title)
    step.tasks.edit_task.fill_form(updated_context.create_task_data)

    support.mock.api
         .update_task(updated_context.id, updated_response)
         .get_tasks([updated_response])
         .get_task(updated_response[:id], updated_response)
         .get_is_valid(updated_response[:id], true)

    step.tasks.edit_task.submit_form
    step.tasks.open_task_info_form(updated_context.title)

    # then
    validate.task.data(updated_context.create_task_data)
  end
end
